File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/NaverMapControlHandler.swift
import NMapsMap
import Flutter

internal protocol NaverMapControlHandler {
    func updateCamera(cameraUpdate: NMFCameraUpdate, onSuccess: @escaping (_ isCanceled: Bool) -> Void)

    func cancelTransitions(reason: Int, onSuccess: @escaping (Any?) -> Void)

    func getCameraPosition(onSuccess: @escaping (_ cameraPosition: Dictionary<String, Any>) -> Void)

    func getContentBounds(withPadding: Bool, onSuccess: @escaping (_ latLngBounds: Dictionary<String, Any>) -> Void)

    func getContentRegion(withPadding: Bool, onSuccess: @escaping (_ latLngs: Array<Dictionary<String, Any>>) -> Void)

    func getLocationOverlay(onSuccess: @escaping (Any?) -> Void)

    func screenLocationToLatLng(nPoint: NPoint, onSuccess: @escaping (_ latLng: Dictionary<String, Any>) -> Void)

    func latLngToScreenLocation(latLng: NMGLatLng, onSuccess: @escaping (_ nPoint: Dictionary<String, Any>) -> Void)

    func getMeterPerDp(onSuccess: @escaping (_ meterPerDp: Double) -> Void)
    
    func getMeterPerDp(lat: Double, zoom: Double, onSuccess: @escaping (_ meterPerDp: Double) -> Void)

    func pickAll(
            nPoint: NPoint,
            dpRadius: Double,
            onSuccess: @escaping (_ pickables: Array<Dictionary<String, Any?>>) -> Void
    )

    func takeSnapshot(
            showControls: Bool,
            compressQuality: Int,
            onSuccess: @escaping (String) -> Void,
            onFailure: @escaping (_ e: FlutterError) -> Void
    )

    func setLocationTrackingMode(locationTrackingMode: NMFMyPositionMode, onSuccess: @escaping (Any?) -> Void)

    func getLocationTrackingMode(onSuccess: @escaping (String) -> Void)

    func addOverlayAll(rawOverlays: Array<Dictionary<String, Any>>, onSuccess: @escaping (Any?) -> Void)

    func deleteOverlay(overlayInfo: NOverlayInfo, onSuccess: @escaping (Any?) -> Void)

    func clearOverlays(type: NOverlayType?, onSuccess: @escaping (Any?) -> Void)
    
    func forceRefresh(onSuccess: @escaping (Any?) -> Void)

    func updateOptions(options: Dictionary<String, Any?>, onSuccess: @escaping (Any?) -> Void)
    
    func updateClusteringOptions(rawOptions: Dictionary<String, Any>, onSuccess: @escaping (Any?) -> Void)
    
    func openMapOpenSourceLicense(onSuccess: @escaping (Any?) -> Void)
    
    func openLegend(onSuccess: @escaping (Any?) -> Void)
    
    func openLegalNotice(onSuccess: @escaping (Any?) -> Void)
}

internal extension  NaverMapControlHandler {
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "updateCamera": updateCamera(
                cameraUpdate: try! NCameraUpdate.fromMessageable(call.arguments!).cameraUpdate, onSuccess: result
        )
        case "cancelTransitions": cancelTransitions(reason: asInt(call.arguments!), onSuccess: result)
        case "getCameraPosition": getCameraPosition(onSuccess: result)
        case "getContentBounds": getContentBounds(withPadding: asBool(call.arguments!), onSuccess: result)
        case "getContentRegion": getContentRegion(withPadding: asBool(call.arguments!), onSuccess: result)
        case "getLocationOverlay": getLocationOverlay(onSuccess: result)
        case "screenLocationToLatLng": screenLocationToLatLng(nPoint: NPoint.fromMessageable(call.arguments!), onSuccess: result)
        case "latLngToScreenLocation": latLngToScreenLocation(latLng: asLatLng(call.arguments!), onSuccess: result)
        case "getMeterPerDp":
            let d = call.arguments as? Dictionary<String, Any?>
            guard
                let latitude = d?["latitude"] as? Double,
                let zoom = d?["zoom"] as? Double
            else {
                getMeterPerDp(onSuccess: result)
                break;
            }

            getMeterPerDp(lat: latitude, zoom: zoom, onSuccess: result)
        case "getMeterPerDpAtLatitude":
            let d = asDict(call.arguments!)
            getMeterPerDp(lat: asDouble(d["latitude"]!), zoom: asDouble(d["zoom"]!), onSuccess: result)
        case "pickAll":
            let d = asDict(call.arguments!)
            pickAll(nPoint: NPoint.fromMessageable(d["point"]!), dpRadius: asDouble(d["radius"]!), onSuccess: result)
        case "takeSnapshot":
            let d = asDict(call.arguments!)
            takeSnapshot(showControls: asBool(d["showControls"]!),
                    compressQuality: asInt(d["compressQuality"]!),
                    onSuccess: result,
                    onFailure: result)
        case "setLocationTrackingMode": setLocationTrackingMode(locationTrackingMode: asLocationTrackingMode(call.arguments!), onSuccess: result)
        case "getLocationTrackingMode": getLocationTrackingMode(onSuccess: result)
        case "addOverlayAll": addOverlayAll(rawOverlays: asArr(call.arguments!, elementCaster: asDict), onSuccess: result)
        case "deleteOverlay": deleteOverlay(overlayInfo: NOverlayInfo.fromMessageable(call.arguments!), onSuccess: result)
        case "clearOverlays": clearOverlays(type: castOrNull(call.arguments, caster: { NOverlayType(rawValue: asString($0))! }), onSuccess: result)
        case "forceRefresh": forceRefresh(onSuccess: result)
        case "updateOptions": updateOptions(options: asNullableDict(call.arguments!), onSuccess: result)
        case "updateClusteringOptions": updateClusteringOptions(rawOptions: asDict(call.arguments!), onSuccess: result)
        case "openMapOpenSourceLicense": openMapOpenSourceLicense(onSuccess: result)
        case "openLegend": openLegend(onSuccess: result)
        case "openLegalNotice": openLegalNotice(onSuccess: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/NaverMapControlHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller

import com.naver.maps.geometry.LatLng
import com.naver.maps.map.CameraUpdate
import com.naver.maps.map.LocationTrackingMode
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asBoolean
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLocationTrackingMode
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.base.NPoint
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.NCameraUpdate
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private fun MethodChannel.Result.send(result: Any? = null) = success(result)

private fun MethodChannel.Result.fail(e: Exception) =
    error(e.message ?: "unknownError", e.stackTrace.toString(), null)

internal interface NaverMapControlHandler {
    fun handle(call: MethodCall, result: MethodChannel.Result) = when (call.method) {
        "updateCamera" -> updateCamera(
            cameraUpdate = call.arguments.let(NCameraUpdate::fromMessageable).toCameraUpdate(),
            onSuccess = result::send,
        )
        "cancelTransitions" -> cancelTransitions(
            reason = call.arguments.asInt(), onSuccess = result::send
        )
        "getCameraPosition" -> getCameraPosition(result::send)
        "getContentBounds" -> getContentBounds(
            withPadding = call.arguments.asBoolean(),
            onSuccess = result::send,
        )
        "getContentRegion" -> getContentRegion(
            withPadding = call.arguments.asBoolean(),
            onSuccess = result::send,
        )
        "getLocationOverlay" -> getLocationOverlay(result::send)
        "screenLocationToLatLng" -> screenLocationToLatLng(
            nPoint = NPoint.fromMessageable(call.arguments),
            onSuccess = result::send,
        )
        "latLngToScreenLocation" -> latLngToScreenLocation(
            latLng = call.arguments.asLatLng(),
            onSuccess = result::send,
        )
        "getMeterPerDp" -> {
            val arguments = call.arguments?.asMap()
            val latitude = arguments?.get("latitude")?.asDouble()
            val zoom = arguments?.get("zoom")?.asDouble()
            if (latitude != null && zoom != null) {
                getMeterPerDp(
                        lat = latitude,
                        zoom = zoom,
                        onSuccess = result::send,
                )
            } else {
                getMeterPerDp(result::send)
            }
        }
        "getMeterPerDpAtLatitude" -> call.arguments.asMap().let {
            getMeterPerDp(
                    lat = it["latitude"]!!.asDouble(),
                    zoom = it["zoom"]!!.asDouble(),
                    onSuccess = result::send,
            )
        }
        "pickAll" -> call.arguments.asMap().let {
            pickAll(
                nPoint = NPoint.fromMessageable(it["point"]!!),
                dpRadius = it["radius"]!!.asDouble(),
                onSuccess = result::send,
            )
        }
        "takeSnapshot" -> call.arguments.asMap().let {
            takeSnapshot(
                showControls = it["showControls"]!!.asBoolean(),
                compressQuality = it["compressQuality"]!!.asInt(),
                onSuccess = result::send,
                onFailure = result::fail,
            )
        }
        "setLocationTrackingMode" -> setLocationTrackingMode(
            locationTrackingMode = call.arguments.asLocationTrackingMode(),
            onSuccess = result::send,
        )
        "getLocationTrackingMode" -> getLocationTrackingMode(result::send)
        "addOverlayAll" -> addOverlayAll(
            rawOverlays = call.arguments.asList { it.asMap() },
            onSuccess = result::send,
        )
        "deleteOverlay" -> deleteOverlay(
            overlayInfo = NOverlayInfo.fromMessageable(call.arguments),
            onSuccess = result::send,
        )
        "clearOverlays" -> clearOverlays(
            type = call.arguments?.toString()?.let(NOverlayType::fromString),
            onSuccess = result::send
        )
        "forceRefresh" -> forceRefresh(onSuccess = result::send)
        "updateOptions" -> updateOptions(
            rawOptions = call.arguments.asMap(), onSuccess = result::send,
        )
        "updateClusteringOptions" -> updateClusteringOptions(
            rawOptions = call.arguments.asMap(), onSuccess = result::send,
        )
        "openMapOpenSourceLicense" -> openMapOpenSourceLicense(result::send)
        "openLegend" -> openLegend(result::send)
        "openLegalNotice" -> openLegalNotice(result::send)
        else -> result.notImplemented()
    }

    fun updateCamera(cameraUpdate: CameraUpdate, onSuccess: (isCanceled: Boolean) -> Unit)

    fun cancelTransitions(reason: Int, onSuccess: () -> Unit)

    fun getCameraPosition(onSuccess: (cameraPosition: Map<String, Any>) -> Unit)

    fun getContentBounds(withPadding: Boolean, onSuccess: (latLngBounds: Map<String, Any>) -> Unit)

    fun getContentRegion(withPadding: Boolean, onSuccess: (latLngs: List<Map<String, Any>>) -> Unit)

    fun getLocationOverlay(onSuccess: (Map<String, Any?>) -> Unit)

    fun screenLocationToLatLng(nPoint: NPoint, onSuccess: (latLng: Map<String, Any>) -> Unit)

    fun latLngToScreenLocation(latLng: LatLng, onSuccess: (nPoint: Map<String, Any>) -> Unit)

    fun getMeterPerDp(onSuccess: (meterPerDp: Double) -> Unit)

    fun getMeterPerDp(lat: Double, zoom: Double, onSuccess: (meterPerDp: Double) -> Unit)

    fun pickAll(
        nPoint: NPoint,
        dpRadius: Double,
        onSuccess: (pickables: List<Map<String, Any?>>) -> Unit,
    )

    fun takeSnapshot(
        showControls: Boolean,
        compressQuality: Int,
        onSuccess: (String) -> Unit, onFailure: (e: Exception) -> Unit,
    )

    fun setLocationTrackingMode(locationTrackingMode: LocationTrackingMode, onSuccess: () -> Unit)

    fun getLocationTrackingMode(onSuccess: (String) -> Unit)

    fun addOverlayAll(rawOverlays: List<Map<String, Any>>, onSuccess: () -> Unit)

    fun deleteOverlay(overlayInfo: NOverlayInfo, onSuccess: () -> Unit)

    fun clearOverlays(type: NOverlayType?, onSuccess: () -> Unit)

    fun forceRefresh(onSuccess: () -> Unit)

    fun updateOptions(rawOptions: Map<String, Any?>, onSuccess: () -> Unit)

    fun updateClusteringOptions(rawOptions: Map<String, Any>, onSuccess: () -> Unit)

    fun openMapOpenSourceLicense(onSuccess: () -> Unit)

    fun openLegend(onSuccess: () -> Unit)

    fun openLegalNotice(onSuccess: () -> Unit)
}